tests/e2e/e2e_test.go

```diff
@@ -1,63 +1,166 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
  */
 
 package e2e
 
 import (
 	"context"
-	"flag"
+	"fmt"
+	"os"
+	"path/filepath"
+	"runtime"
+	"strconv"
 	"testing"
 
 	. "github.com/onsi/ginkgo/v2"
 	. "github.com/onsi/gomega"
 )
 
+const (
+	vectorAddImage   = "nvcr.io/nvidia/k8s/cuda-sample:vectoradd-cuda12.5.0"
+	deviceQueryImage = "nvcr.io/nvidia/k8s/cuda-sample:devicequery-cuda12.5.0"
+	cudaImage        = "nvcr.io/nvidia/cuda:12.8.0-base-ubi8"
+)
+
 // Test context
 var (
 	ctx context.Context
 
 	installCTK bool
 
-	image string
+	ImageRepo string
+	ImageTag  string
 
-	sshKey  string
-	sshUser string
-	host    string
-	sshPort string
-)
+	sshKey      string
+	sshUser     string
+	host        string
+	sshPort     string
+	cwd         string
+	packagePath string
 
-func init() {
-	flag.BoolVar(&installCTK, "install-ctk", false, "Install the NVIDIA Container Toolkit")
-	flag.StringVar(&image, "toolkit-image", "", "Repository of the image to test")
-	flag.StringVar(&sshKey, "ssh-key", "", "SSH key to use for remote login")
-	flag.StringVar(&sshUser, "ssh-user", "", "SSH user to use for remote login")
-	flag.StringVar(&host, "remote-host", "", "Hostname of the remote machine")
-	flag.StringVar(&sshPort, "remote-port", "22", "SSH port to use for remote login")
-}
+	runner Runner
+)
 
 func TestMain(t *testing.T) {
-	suiteName := "NVIDIA Container Toolkit E2E"
+	suiteName := "E2E NVIDIA Container Toolkit"
 
 	RegisterFailHandler(Fail)
+
+	ctx = context.Background()
+	getTestEnv()
+
 	RunSpecs(t,
 		suiteName,
 	)
 }
 
 // BeforeSuite runs before the test suite
 var _ = BeforeSuite(func() {
-	ctx = context.Background()
+	runner = NewRunner(
+		WithHost(host),
+		WithPort(sshPort),
+		WithSshKey(sshKey),
+		WithSshUser(sshUser),
+	)
+
+	if installCTK {
+		installer, err := NewToolkitInstaller(
+			WithRunner(runner),
+			WithImage(ImageRepo+":"+ImageTag),
+			WithTemplate(dockerInstallTemplate),
+		)
+		Expect(err).ToNot(HaveOccurred())
+
+		err = installer.Install()
+		Expect(err).ToNot(HaveOccurred())
+	}
+
+	_, _, err := runner.Run("docker pull ubuntu")
+	Expect(err).ToNot(HaveOccurred())
+
+	_, _, err = runner.Run(fmt.Sprintf("docker pull %s", vectorAddImage))
+	Expect(err).ToNot(HaveOccurred())
+
+	_, _, err = runner.Run(fmt.Sprintf("docker pull %s", deviceQueryImage))
+	Expect(err).ToNot(HaveOccurred())
+	_, _, err = runner.Run(fmt.Sprintf("docker pull %s", cudaImage))
+	Expect(err).ToNot(HaveOccurred())
```

Comment:
What if a user is only running a subset of tests? The pull of the images should be LOCAL to the tests that use the images. Do you have a motivation for collecting them in one place?

Reply:
Maybe as a general comment: When a developer is writing tests, we want to avoid having them have to remember that they also have to add the pull checks here when adding tests for new functionality.

Reply:
Fixed